### DISCUSSION TITLE
ServiceProvider's bootWhen method with serviceProvider arg always being undefined

### DISCUSSION THREAD

**Discussion by c0nsol3** [ArkEcosystem/core#4524] [Q&A] [answered]
I wish my plugin to boot after a specific serviceProvider, but serviceProvider is always `undefined`.
[This](https://github.com/deadlock-delegate/notifier/blob/master/src/service-provider.ts#L23-L25) is the bootWhen that I do in my plugin, which always fails on mainnet (used installer to setup).

Code sample:
```ts
export class ServiceProvider extends Providers.ServiceProvider {
    @Container.inject(Container.Identifiers.LogService)
    private readonly logger!: Contracts.Kernel.Logger;

    public async register(): Promise<void> {
        // this gets called when core starts
    }

    public async boot(): Promise<void> {
    }

    public async bootWhen(serviceProvider?: string): Promise<boolean> {
        // serviceProvider is always `undefined`
        return !!this.config().get("enabled") && serviceProvider === "@arkecosystem/core-blockchain";
    }

    public async dispose(): Promise<void> {
    }
}
```

Strange thing is I didn't notice this issue when installing from source and running testnet on my local computer.

**Comment by KovacZan** [comment]
Can  you try adding `index.ts `file and add `export * from "./service-provider";`
It think it should work then.

**Reply by KovacZan** [reply]
Ah sorry wrong anwser.

**Comment by sebastijankuzner** [comment, marked as answer]
I suspect that you changed the plugin order in app.json file. Can you check that?

Generally you don't need to wait to specific serviceProvider to be booted.  The boot process will boot them in the same order as they are provided in app.json. 

It is possible that you located your plugin after the `@arkecosystem/core-blockchain`and this mean that core-blockchain is booted before the handler for your plugin is registered, which results in a missing event, where serviceProvider variable should be equal `@arkecosystem/core-blockchain`. 


Approach when your plugin depends on another serviceProvider is recommended when dependent plugin is disposed and then booted again, but not on the initial boot.

**Reply by c0nsol3** [reply]
I see. Yes, `@deadlock-delegate/notifier` is the very last plugin in the `app.json`.  I thought order of packages in `app.json` no longer matters if `bootWhen` is defined.

Is there actually a case where plugin is booted after initial boot?

**Reply by faustbrian** [reply]
The use-cases for `bootWhen` and `disposeWhen` are plugins that should either be temporarily enabled (like maintenance or repair plugins) or plugins that want to activate themselves only after a certain time or milestone. If your only need is to have it booted after another plugin then simply list it after the plugin it depends on or as last.
